migrations/dynamodb-migrations/2021-08-04T10:06:50Z-migrate-sampleids-to-experiment.py

```diff
@@ -0,0 +1,156 @@
+    sample_ids = [sample_id for sample_id in samples['M'] if sample_id != 'ids']
+
+    sample_names = []
+
+    # Sample names
+    for sample_id in sample_ids:
+        sample_names.append(samples['M'][sample_id]['M']['name']['S'])
+
+    # Sample IDS and names are sorted to be invariant
+    sample_ids.sort()
+    sample_names.sort()
+
+    input_type = experiment["meta"]['M']["type"].get('S', '10x')
+
+    hash_params = {
+        "organism": organism,
+        "input": {"type" : input_type},
+        "sampleIds": sample_ids,
+        "sampleNames": sample_names,
+    }
+
+    metadata_values = OrderedDict()
+
+    if len(project['M']['metadataKeys']['L']) > 0:
+
+        metadata_keys = [metadata['S'] for metadata in project['M']['metadataKeys']['L']]
+
+        for key in metadata_keys:
+            # Replace '-' in key to '_'if
+            sanitizedKey = key.replace('-', '_')
+
+            for sample_id in sample_ids:
+
+                metadata_value = "N.A." # default metadata value
+
+                if samples['M'][sample_id]['M']['metadata']['M'].get(key):
+                    metadata_value = samples['M'][sample_id]['M']['metadata']['M'][key]['S']
+
+                if not metadata_values.get(sanitizedKey):
+                    metadata_values[sanitizedKey] = []    
+
+                metadata_values[sanitizedKey].append(metadata_value)
+
+        hash_params['metadata'] = metadata_values
+
+    hash_params_string = json.dumps(hash_params).replace(", ", ",").replace(": ", ":").encode('utf-8')
+
+    return hashlib.sha1(hash_params_string).hexdigest()
+
+
+print(f"=== Doing migration for PR 408 ===")
+
+for experiment in experiments:
+
+    try:
+        experiment_id = experiment['experimentId']
+        print(f"\n=========")
+        print(f"Preparing migration for experiment {experiment_id['S']}")
+
+        if not experiment.get('createdAt') : print('Project does not contain createdAt, inserting current timestamp')
+        created_date = experiment['createdAt'] if experiment.get('createdAt') else { "S" : datetime.now().isoformat() }
+
+        if not experiment.get('projectId') : raise Exception('Experiment does not have associated project')
+        project_id = experiment['projectId']
+
+        project = client.get_item(
+            TableName=projects_table,
+            Key={"projectUuid" : project_id },
+            ProjectionExpression="projects"
+        )
+
+        if not project.get('Item') : raise Exception('Project does not exist')
+        samples = project['Item']['projects']['M']['samples']
+        
+        if not samples.get('L') : raise Exception('Project does not contain samples')
+
+        # Get variables
+
+        print("calculating new GEM2S hash...")
+        experiment_record = client.get_item(
+            TableName=experiments_table,
+            Key={"experimentId": experiment_id }
+        )['Item']
+
+        project_record = client.get_item(
```

Comment:
you can use previous `project` variable instead of querying it again

Reply:
Removed